casper/src/main/rholang/MakeMint.rho

```diff
@@ -0,0 +1,60 @@
+//scalapackage coop.rchain.rholang.mint
+
+//based on http://erights.org/elib/capability/ode/ode-capabilities.html#simple-money
+//requires NonNegativeNumber, MakeBrandPair
+contract @["MakeMint", "int2NN"](@x, return) = {
+  new nnCh in {
+    @"NonNegativeNumber"!(x, *nnCh) | for(@nn <- nnCh) {
+      return!(nn)
+    }
+  }
+} |
+contract @"MakeMint"(return) = {
+  new pairCh, thisMint, internalMakePurse in {
+    @"MakeBrandPair"!(*pairCh) | for(@p <- pairCh) {
+      match p {
+        [seal, unseal] => {
+          contract @[*thisMint, "makePurse"](@init, return) = {
+            new balanceCh in {
+              @["MakeMint", "int2NN"]!(init, *balanceCh) | for(@balance <- balanceCh) {
+                internalMakePurse!(balance, *return)
+              }
+            }
+          } |
+          contract internalMakePurse(@balance, return) = {
+            //balance must be a name that NonNegativeNumber contract is listening on
+            new thisPurse in {
+              contract @[*thisPurse, "decr"](@amount, success) = { @[balance, "sub"]!(amount, *success) } |
+              contract @[*thisPurse, "getBalance"](return) = { @[balance, "value"]!(*return) } |
+              contract @[*thisPurse, "sprout"](return) = { @[*thisMint, "makePurse"]!(0, *return) } |
+              contract @[*thisPurse, "getDecr"](return) = { @seal!([*thisPurse, "decr"], *return) } |
+              contract @[*thisPurse, "deposit"](@amount, @src, success) = {
+                new result in {
+                  @[src, "getDecr"]!(*result) | for(@sealedSrcDecr <- result) { //call getDecr function in src
```

Comment:
There's a simpler way to do this without the sealer unsealer. Because we have bundles, we can use unforgeable names in a challenge response protocol.

Mint has a new name as a tag.

getDecr instead of giving a sealed decr would give a new name as a challenge, and an answer channel. If bundle0{*tag | *challenge} is received on the answer channel, decr can be passed.